File: 지찬/프로그래머스 코테 kit/올바른괄호.java
//https://school.programmers.co.kr/learn/courses/30/lessons/12909
import java.util.Stack;

class Solution {
    boolean solution(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c :
                s.toCharArray()) {
            if (c == '(') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                stack.pop();
            }
        }

        if (!stack.isEmpty()) {
            return false;
        }


        return true;
    }

    public static void main(String[] args) {
        String s = "()()";
        Solution solution = new Solution();
        System.out.println(solution.solution(s));
    }
}

File: 종우/올바른괄호.py
# 1006_3
# 올바른 괄호
# https://school.programmers.co.kr/learn/courses/30/lessons/12909

from collections import deque

def solution(s):
    stack = deque()
    for i in s:
        if i == "(":
            stack.appendleft(i)
        else:
            try:
                if not(stack.popleft() == "("):
                    stack.leftappend(i)
            except IndexError:
                return False
    if len(stack) == 0:
        return True
    return False